lib/puppet/state_machine.rb

```diff
@@ -0,0 +1,87 @@
+module Puppet
+  # Define state machines for modeling complex operations.
+  #
+  # A given state machine is immutable once built; invoking the state machine generates a context
+  # for that invocation that leaves the state machine and associated states unmodified.
+  class StateMachine
+    require 'puppet/state_machine/state'
+    require 'puppet/state_machine/machine_builder'
+
+    # @return [Puppet::StateMachine]
+    def self.build(name, &block)
+      MachineBuilder.new(name).build(&block)
+    end
+
+    def self.compose(name, machines, &block)
+      MachineBuilder.new(name).compose(machines, &block)
+    end
+
+    # @!attribute [r] name
+    #   @return [String] The description of this state machine
+    attr_reader :name
+
+    # @!attribute [r] states
+    #   @api private
+    attr_reader :states
+
+    # @param name [String] The description of this state machine.
+    # @param states [Hash<Symbol, Puppet::StateMachine::State>]
+    # @param start_state [Symbol] The name of the initial state used when executing this machine.
+    def initialize(name, states, start_state)
+      @name = name
+      @states = states
+      @start_state = start_state
+    end
+
+    # @return [Puppet::StateMachine::State]
+    def start_state
+      @states[@start_state]
+    end
+
+    def state(state_name)
+      @states[state_name]
+    end
+
+    # Generate a set of state machines with the given namespace added to the state name
+    # and transitions.
+    #
+    # @param ns [Symbol]
+    # @return [Hash<(Symbol, Array<Symbol>), Puppet::StateMachine::State>]
+    def namespaced_states(ns)
+      @states.values.inject({}) do |hash, state|
+        namespaced = state.namespace(ns)
+        hash[namespaced.name] = namespaced
+        hash
+      end
+    end
+
+    # Generate a new state machine context and run this machine.
+    def call
+      result = :running
+      current = start_state
+      loop do
+        result = current.action
+
+        if current.error?
+          return :errored
+        elsif current.final?
+          return :complete
+        end
```

Comment:
@hlindberg my original approach has been to model this somewhat like a DFA that takes some input and runs till completion or error. In this case we don't have an input string and the transitions are determined by the current state, but the model was to run a state machine to completion. One option for this is to extract the running logic to a separate class again, so if we want to model different types of behavior or have an async machine we can provide a different runner class that implements the desired behavior. What do you think of this approach?